File: retaRust/python_arch_reference/reta_architecture/completion_runtime.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, MutableMapping, Optional, Sequence, Set, Tuple

from .facade import RetaArchitecture
from .prompt_runtime import PromptRuntimeBundle


def sort_completion_key(key: str, i18n) -> Tuple[int, str]:
    key = str(key)
    if not key.startswith("1") and not any((key[:3] == a[:3] for a in i18n.EIGS_N_R)):
        if key in (
            i18n.befehle2["absicht"],
            i18n.befehle2["hilfe"],
            i18n.befehle2["kurzbefehle"],
        ) and len(key) != 1:
            return (0, key)
        elif key in (
            i18n.befehle2["universum"],
            i18n.befehle2["thomas"],
            i18n.befehle2["befehle"],
            i18n.befehle2["groesse"],
        ) and len(key) != 1:
            return (1, key)
        elif key in (
            i18n.befehle2["reta"],
            i18n.befehle2["bewusstsein"],
            i18n.befehle2["geist"],
            i18n.befehle2["emotion"],
            i18n.befehle2["impulse"],
        ) and len(key) != 1:
            return (2, key)
        elif key in (
            i18n.befehle2["loggen"],
            i18n.befehle2["nichtloggen"],
            i18n.befehle2["exit"],
            i18n.befehle2["quit"],
        ) and len(key) != 1:
            return (3, key)
        elif len({key} & set(i18n.befehle2.values())) > 0 and not (
            key[:3] not in ("15_", "16_", i18n.EIGS_N_R[1][:3]),
            i18n.EIGS_N_R[0][:3],
        ) and len(key) != 1:
            return (4, key)
        elif key in i18n.befehle and len(key) != 1:
            return (5, key)
        elif not key.startswith("1") and len(key) != 1:
            return (6, key)
        elif len(key.strip()) == 1:
            return (7, key)
        elif key[:3] == "15_" or key == "15":
            return (8, key)
        elif key[:3] == "16_" or key == "16":
            return (9, key)
        elif key.startswith("1") and len(key) != 1:
            if key.startswith("15"):
                return (10, key)
            elif key.startswith("16"):
                return (11, key)
            else:
                return (11, key)
        elif any((key[:3] == a[:3] for a in i18n.EIGS_N_R)):
            return (19, key)
        else:
            return (13, key)
    elif any((key[:3] == a[:3] for a in i18n.EIGS_N_R)):
        return (19, key)
    else:
        return (14, key)


@dataclass
class CompletionRuntimeBundle:
    prompt_runtime: PromptRuntimeBundle
    befehle: List[str]
    befehle2: Set[str]
    befehle2_list: List[str]
    haupt_for_neben: List[str]
    haupt_for_neben_set: Set[str]
    ausgabe_art: List[str]
    ausgabe_paras: List[str]
    kombi_main_paras: List[str]
    main_parameters: List[str]
    spalten: List[str]
    spalten_dict: Dict[str, List[str]]
    zeilen_paras: List[str]
    zeilen_typen: List[str]
    zeilen_zeit: List[str]
    zeilen_typen_b: List[str]
    kombi_value_options: Dict[str, List[str]] = field(default_factory=dict)

    @property
    def program(self):
        return self.prompt_runtime.program

    @property
    def vocabulary(self):
        return self.prompt_runtime.vocabulary

    def start_commands(self, include_numeric_shortcuts: bool = False) -> List[str]:
        commands = list(self.befehle)
        if include_numeric_shortcuts:
            for item in ("15_", "16_"):
                if item not in commands:
                    commands.append(item)
        return commands

    def snapshot(self) -> Dict[str, object]:
        return {
            "befehle_len": len(self.befehle),
            "befehle2_len": len(self.befehle2),
            "haupt_for_neben_len": len(self.haupt_for_neben),
            "spalten_dict_keys": len(self.spalten_dict),
            "kombi_option_keys": sorted(self.kombi_value_options.keys()),
            "start_commands_with_numeric_shortcuts": self.start_commands(
                include_numeric_shortcuts=True
            )[:10],
            "prompt_runtime": self.prompt_runtime.snapshot(),
        }


class CompletionRuntimeBuilder:
    def __init__(self, architecture: RetaArchitecture, i18n) -> None:
        self.architecture = architecture
        self.i18n = i18n

    def build(self) -> CompletionRuntimeBundle:
        prompt_runtime = self.architecture.bootstrap_prompt_runtime(i18n=self.i18n)
        vocabulary = prompt_runtime.vocabulary
        program = prompt_runtime.program
        befehle2_list = list(
            sorted(set(vocabulary.befehle2), key=lambda item: sort_completion_key(item, self.i18n))
        )
        kombi_galaxie = [
            item
            for sublist in program.kombiParaNdataMatrix.values()
            for item in sublist
        ]
        kombi_universum = [
            item
            for sublist in program.kombiParaNdataMatrix2.values()
            for item in sublist
        ]
        return CompletionRuntimeBundle(
            prompt_runtime=prompt_runtime,
            befehle=list(sorted(vocabulary.befehle, key=lambda item: sort_completion_key(item, self.i18n))),
            befehle2=set(befehle2_list),
            befehle2_list=befehle2_list,
            haupt_for_neben=list(vocabulary.haupt_for_neben),
            haupt_for_neben_set=set(vocabulary.haupt_for_neben_set),
            ausgabe_art=list(vocabulary.ausgabe_art),
            ausgabe_paras=list(vocabulary.ausgabe_paras),
            kombi_main_paras=list(vocabulary.kombi_main_paras),
            main_parameters=list(vocabulary.main_parameters),
            spalten=list(vocabulary.spalten),
            spalten_dict={key: list(value) for key, value in vocabulary.spalten_dict.items()},
            zeilen_paras=list(vocabulary.zeilen_paras),
            zeilen_typen=list(vocabulary.zeilen_typen),
            zeilen_zeit=list(vocabulary.zeilen_zeit),
            zeilen_typen_b=list(vocabulary.zeilen_typen_b),
            kombi_value_options={
                self.i18n.kombiMainParas["galaxie"]: kombi_galaxie,
                self.i18n.kombiMainParas["universum"]: kombi_universum,
                "*": kombi_galaxie + kombi_universum,
            },
        )


_COMPLETION_RUNTIME_CACHE: MutableMapping[Tuple[str, str], CompletionRuntimeBundle] = {}


def bootstrap_completion_runtime(
    repo_root: Optional[Path] = None,
    architecture: Optional[RetaArchitecture] = None,
    i18n=None,
    force_rebuild: bool = False,
) -> CompletionRuntimeBundle:
    architecture = architecture or RetaArchitecture.bootstrap(repo_root)
    if i18n is None:
        import i18n.words_runtime as i18n  # noqa: WPS433

    cache_key = (
        str(architecture.repo_root.resolve()),
        str(getattr(i18n, "__name__", type(i18n).__name__)),
    )
    if force_rebuild or cache_key not in _COMPLETION_RUNTIME_CACHE:
        _COMPLETION_RUNTIME_CACHE[cache_key] = CompletionRuntimeBuilder(
            architecture, i18n
        ).build()
    return _COMPLETION_RUNTIME_CACHE[cache_key]

File: retaRust/crates/reta_architecture/src/completion_runtime.rs
//! Completion runtime vocabulary skeleton transcompiled from
//! `python_arch_reference/reta_architecture/completion_runtime.py`.
//!
//! The Python builder still extracts the full i18n vocabulary.  This Rust
//! module owns the typed completion bundle shape and deterministic sort key so
//! prompt frontends can move onto Rust one section at a time.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct CompletionSortKey {
    pub group: u8,
    pub key: String,
}

pub fn sort_completion_key(key: &str) -> CompletionSortKey {
    let trimmed = key.trim().to_string();
    let group = if trimmed.len() == 1 {
        7
    } else if trimmed == "15" || trimmed.starts_with("15_") {
        8
    } else if trimmed == "16" || trimmed.starts_with("16_") {
        9
    } else if trimmed.starts_with('1') {
        11
    } else if matches!(trimmed.as_str(), "hilfe" | "help" | "kurzbefehle" | "absicht") {
        0
    } else if matches!(trimmed.as_str(), "universum" | "thomas" | "befehle" | "groesse") {
        1
    } else if matches!(trimmed.as_str(), "reta" | "bewusstsein" | "geist" | "emotion" | "impulse") {
        2
    } else if matches!(trimmed.as_str(), "loggen" | "nichtloggen" | "exit" | "quit") {
        3
    } else {
        6
    };
    CompletionSortKey { group, key: trimmed }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct CompletionRuntimeBundle {
    pub befehle: Vec<String>,
    pub befehle2: BTreeSet<String>,
    pub befehle2_list: Vec<String>,
    pub haupt_for_neben: Vec<String>,
    pub haupt_for_neben_set: BTreeSet<String>,
    pub ausgabe_art: Vec<String>,
    pub ausgabe_paras: Vec<String>,
    pub kombi_main_paras: Vec<String>,
    pub main_parameters: Vec<String>,
    pub spalten: Vec<String>,
    pub spalten_dict: BTreeMap<String, Vec<String>>,
    pub zeilen_paras: Vec<String>,
    pub zeilen_typen: Vec<String>,
    pub zeilen_zeit: Vec<String>,
    pub zeilen_typen_b: Vec<String>,
    pub kombi_value_options: BTreeMap<String, Vec<String>>,
}

impl Default for CompletionRuntimeBundle {
    fn default() -> Self {
        let mut befehle = vec![
            "hilfe".to_string(),
            "reta".to_string(),
            "exit".to_string(),
            "quit".to_string(),
            "15_".to_string(),
            "16_".to_string(),
        ];
        befehle.sort_by_key(|item| sort_completion_key(item));
        let befehle2_list = befehle.clone();
        Self {
            befehle: befehle.clone(),
            befehle2: befehle.iter().cloned().collect(),
            befehle2_list,
            haupt_for_neben: vec!["zeilen".to_string(), "spalten".to_string(), "ausgabe".to_string()],
            haupt_for_neben_set: BTreeSet::from(["zeilen".to_string(), "spalten".to_string(), "ausgabe".to_string()]),
            ausgabe_art: vec!["bbcode".to_string(), "html".to_string(), "csv".to_string(), "shell".to_string(), "markdown".to_string(), "emacs".to_string(), "nichts".to_string()],
            ausgabe_paras: vec!["art".to_string(), "breite".to_string(), "breiten".to_string()],
            kombi_main_paras: vec!["galaxie".to_string(), "universum".to_string()],
            main_parameters: vec!["zeilen".to_string(), "spalten".to_string(), "kombination".to_string(), "ausgabe".to_string()],
            spalten: Vec::new(),
            spalten_dict: BTreeMap::new(),
            zeilen_paras: vec!["alles".to_string(), "vorhervonausschnitt".to_string(), "oberesmaximum".to_string()],
            zeilen_typen: vec!["sonne".to_string(), "planet".to_string(), "mond".to_string()],
            zeilen_zeit: vec!["heute".to_string(), "gestern".to_string(), "morgen".to_string()],
            zeilen_typen_b: Vec::new(),
            kombi_value_options: BTreeMap::new(),
        }
    }
}

impl CompletionRuntimeBundle {
    pub fn start_commands(&self, include_numeric_shortcuts: bool) -> Vec<String> {
        let mut commands = self.befehle.clone();
        if include_numeric_shortcuts {
            for item in ["15_", "16_"] {
                if !commands.iter().any(|existing| existing == item) {
                    commands.push(item.to_string());
                }
            }
        }
        commands.sort_by_key(|item| sort_completion_key(item));
        commands
    }

    pub fn snapshot(&self) -> CompletionRuntimeSnapshot {
        CompletionRuntimeSnapshot {
            befehle_len: self.befehle.len(),
            befehle2_len: self.befehle2.len(),
            haupt_for_neben_len: self.haupt_for_neben.len(),
            spalten_dict_keys: self.spalten_dict.len(),
            kombi_option_keys: self.kombi_value_options.keys().cloned().collect(),
            start_commands_with_numeric_shortcuts: self
                .start_commands(true)
                .into_iter()
                .take(10)
                .collect(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct CompletionRuntimeSnapshot {
    pub befehle_len: usize,
    pub befehle2_len: usize,
    pub haupt_for_neben_len: usize,
    pub spalten_dict_keys: usize,
    pub kombi_option_keys: Vec<String>,
    pub start_commands_with_numeric_shortcuts: Vec<String>,
}

pub fn bootstrap_completion_runtime() -> CompletionRuntimeBundle {
    CompletionRuntimeBundle::default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sort_key_keeps_numeric_shortcuts_late() {
        assert!(sort_completion_key("hilfe") < sort_completion_key("15_"));
        assert_eq!(sort_completion_key("x").group, 7);
    }

    #[test]
    fn bundle_starts_with_prompt_commands() {
        let bundle = bootstrap_completion_runtime();
        assert!(bundle.start_commands(true).contains(&"15_".to_string()));
        assert!(bundle.snapshot().befehle_len > 0);
    }
}


// Stage 15: explicit py-reta-arch compatibility surface markers.
// These markers keep historical Python architecture symbol names visible
// while the Rust implementation is migrated module by module. They are
// not a claim of byte-exact semantic replacement for every listed helper.
#[allow(dead_code)]
pub const PY_ARCH_STAGE15_SURFACE: &[&str] = &[
    "CompletionRuntimeBuilder",
    "__init__",
    "build",
    "program",
    "vocabulary",
];

#[allow(dead_code)]
pub fn stage15_py_surface_names() -> &'static [&'static str] {
    PY_ARCH_STAGE15_SURFACE
}
